### DISCUSSION TITLE
Cancel action with filter hook without returning an exception

### DISCUSSION THREAD

**Discussion by enzolupia** [directus/directus#17695] [Archive]
According to https://docs.directus.io/extensions/hooks.html and our previous experience with filter hooks, the only way to cancel an action is to throw a managed exception (i.e. InvalidPayloadException).
Is it possible to just cancel an action without exception? This will prevent the UI to show the error notice.
For example I want to intercept the save action and in specific circumstances instead of just saving the item I want to save it as a copy.

**Comment by rolodato** [comment]
For any fellow googlers, this is somewhat possible by using the `status` parameter of `createError`. The [example from the docs](https://docs.directus.io/extensions/hooks.html) returns a 500 error:

```js
import { createError } from '@directus/errors';

const InvalidPayloadError = createError('INVALID_PAYLOAD_ERROR', 'Something went wrong...', 500);
```

We can change this to a 200 status code which hides any error messages from the UI:

```js
const CancelledError = createError('CANCELLED_ERROR', 'Nope', 200);
```

This will cause your `POST /your-collection` endpoint (or whatever method you hook into) to return a 200 but with the following response body:

```json
{
  "errors": [
    {
      "message": "Nope",
      "extensions": {
        "code": "CANCELLED_ERROR"
      }
    }
  ]
}
```

I don't know if this is a sustainable approach, i.e. Directus or other API consumers might start treating this as an actual error despite the 200 status code. I'm currently using this approach to build a filter hook for soft-deleting data by archiving it instead.

**Comment by rolodato** [comment]
This is what I've able to quickly come up with - if an item is deleted, the deletion is prevented and the item is archived instead. If the item was already archived, it's deleted for real.

It's not ideal as it overloads the concept of "archived", and it also makes the deletion UI message untrue since it really can be reverted now (in some cases), but it's good enough for my use case. I want to not punish users for accidentally deleting something and then having to bother an IT admin to restore a backup for them, while still letting them hard-delete stuff without needing to change anything in the UI.

```ini
# .env
SOFT_DELETE_COLLECTIONS=json:["my_collection_name"]
```

```ts
// extension's index.ts

import { defineHook } from "@directus/extensions-sdk";
import { createError } from '@directus/errors';

const SoftDeletedError = createError('SOFT_DELETED', 'Soft deletion successful. This is not an actual error.', 200);

export default defineHook(async ({ filter }, { env, logger, database }) => {
    const { SOFT_DELETE_COLLECTIONS } = env;
    logger.info(`defining soft-delete hooks for ${SOFT_DELETE_COLLECTIONS}`);
    for (const c of SOFT_DELETE_COLLECTIONS) {
        filter(`${c}.items.delete`, async (items) => {
            let archivedRows = 0;
            await database.transaction(async tx => {
                // If item is already archived, delete it for real
                await tx(c)
                    .whereIn('id', items as string[])
                    .where('status', 'archived')
                    .del()

                // Archive the remaining items which were not deleted
                archivedRows = await tx(c)
                    .whereIn('id', items as string[])
                    .update({ status: 'archived' })
            })
            // Abort original deletion event but don't show an error in the UI
            // Preserve original behavior if nothing was soft-deleted
            if (archivedRows > 0) {
                logger.info(`succesfully soft-deleted ${archivedRows} ${c} items with ids: ${items}`);
                throw new SoftDeletedError();
            }
        })
    }
});
```

**Comment by Stathis90** [comment]
Does this work for you?
I am trying to create a hook like this but I get error on registering it
```
import { createError } from "@directus/errors";

const InvalidPayloadError = createError(INVALID_PAYLOAD_ERROR", "Something went wrong...", 500);

var index = ({
	action
}) => {
	action("IR_Register.items.update", async (input, ) => {
		try {
			const {
				payload
			} = input;
			// Check if billing_prov or service_prov is present in the payload
			if (payload.billing_prov || payload.service_prov) {
				// Reject the update
				throw new InvalidPayloadError();
			}
		} catch (e) {
			console.error("Error executing custom procedure:", e);
			throw new InvalidPayloadError();
		}
	});

};

export { index as default };
```

WARN: Cannot use import statement outside a module
    err: {
      "type": "SyntaxError",
      "message": "Cannot use import statement outside a module",
      "stack":
          /directus/extensions/hooks/update-hook/index.cjs:1
          import { createError } from '@directus/errors';
          ^^^^^^
          
          SyntaxError: Cannot use import statement outside a module

**Reply by rolodato** [reply]
You're writing a `.cjs` file, which uses CommonJS for modules (Node.js `require` function) and not ES modules (`import` syntax). I'm using TypeScript in my example, which uses ES modules. Make sure to use the correct module syntax depending on how you are writing your extension code.

**Comment by jclaveau** [comment]
I opened a proper feature request for this issue https://github.com/directus/directus/discussions/25411
